File: Product-of-Array-Except-Self.java
// with O(1) Space Complexity
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int pre = 1 ;
        int pos = 1 ; 

        int[] res = new int[nums.length];

        for(int i = 0 ;i<nums.length;i++){
            res[i] = pre ;
            pre *= nums[i] ;
        }
        for(int i = nums.length-1;i>=0;i--){
            res[i] *= pos ;
            pos *= nums[i];
        }
        return res; 
    }
}

File: Product-of-Array-Except-Self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        vector<int> pre(n,1);
        vector<int> pos(n,1);
        vector<int> res(n,1);

        for(int i=1;i<n;i++){
            pre[i] = pre[i-1] * nums[i-1] ; 
        }


        for(int i=n-2;i>=0;i--){
            pos[i] = pos[i+1] * nums[i+1] ;
        }

        for(int i = 0 ;i<n;i++){
            res[i] = pre[i] * pos[i];
        }

        return res;
    }
};